### ISSUE TITLE
Does Bitblas support P40's DP4A for matrix?

### ISSUE THREAD

**Issue by sorasoras** [microsoft/BitBLAS#251]
Same as the title.
Since P40 don't have tensor core but it does have fair  int8 perf via DP4A.

**Comment by LeiWang1999** [comment]
@sorasoras , for gemv case (when M=1) we do support dp4a. GEMM case is not supported currently, thanks for your attention, support high performance dp4a with gemm is simple, we'll soon append this feature :)

**Comment by LeiWang1999** [comment]
hi @sorasoras , dp4a and simt implementation (with dequant support) has been implemented via pr #255 with our experimental `tilelang` backend, to utilize please checkout:

```python
import bitblas
M = N = K = 2048
A_dtype = "int8"
W_dtype = "int8"
accum_dtype = "int32"
out_dtype = "int32"

bitblas.set_log_level("DEBUG")
matmul_config = bitblas.MatmulConfig(
    M=1024,  # M dimension
    N=N,  # N dimension
    K=K,  # K dimension
    A_dtype=A_dtype,  # activation A dtype
    W_dtype=W_dtype,  # weight W dtype
    accum_dtype=accum_dtype,  # accumulation dtype
    out_dtype=out_dtype,  # output dtype
    layout="nt",  # matrix layout, "nt" indicates the layout of A is non-transpose and the layout of W is transpose
    propagate_b=False,  # propagate B matrix
)

matmul = bitblas.Matmul(config=matmul_config, enable_tuning=False, backend="tl")

# matmul.hardware_aware_finetune()

print(matmul.get_source())
latency = matmul.profile_latency()
print(latency)
```


And example generated source:


```cpp
extern "C" __global__ void __launch_bounds__(256) matmul_m1024n2048k2048_i8xi8_default_kernel(signed char* __restrict__ A, signed char* __restrict__ B, int* __restrict__ C) {
  extern __shared__ __align__(1024) uchar buf_dyn_shmem[];
  int C_local[64];
  signed char A_local[128];
  signed char B_local[128];
  for (int i = 0; i < 16; ++i) {
    *(int4*)(C_local + (i * 4)) = make_int4(0, 0, 0, 0);
  }
  for (int ko = 0; ko < 128; ++ko) {
    __syncthreads();
    if (((int)threadIdx.x) < 128) {
      *(int4*)(((signed char*)buf_dyn_shmem) + (((int)threadIdx.x) * 16)) = *(int4*)(A + (((((int)blockIdx.y) * 262144) + (((int)threadIdx.x) * 2048)) + (ko * 16)));
    }
    if (((int)threadIdx.x) < 128) {
      *(int4*)(((signed char*)buf_dyn_shmem) + ((((int)threadIdx.x) * 16) + 2048)) = *(int4*)(B + (((((int)blockIdx.x) * 262144) + (((int)threadIdx.x) * 2048)) + (ko * 16)));
    }
    __syncthreads();
    for (int i_1 = 0; i_1 < 8; ++i_1) {
      *(int4*)(A_local + (i_1 * 16)) = *(int4*)(((signed char*)buf_dyn_shmem) + (((((int)threadIdx.x) & 15) * 128) + (i_1 * 16)));
    }
    for (int i_2 = 0; i_2 < 8; ++i_2) {
      *(int4*)(B_local + (i_2 * 16)) = *(int4*)(((signed char*)buf_dyn_shmem) + ((((((int)threadIdx.x) >> 4) * 128) + (i_2 * 16)) + 2048));
    }
    for (int i_3 = 0; i_3 < 8; ++i_3) {
      for (int j = 0; j < 8; ++j) {
        for (int mk = 0; mk < 4; ++mk) {
          DP4A((&(A_local[((i_3 * 16) + (mk * 4))])), (&(B_local[((j * 16) + (mk * 4))])), (&(C_local[((i_3 * 8) + j)])));
        }
      }
    }
  }
  for (int i_4 = 0; i_4 < 8; ++i_4) {
    for (int j_1 = 0; j_1 < 8; ++j_1) {
      C[((((((((int)blockIdx.y) * 262144) + ((((int)threadIdx.x) & 15) * 16384)) + (i_4 * 2048)) + (((int)blockIdx.x) * 128)) + ((((int)threadIdx.x) >> 4) * 8)) + j_1)] = C_local[((i_4 * 8) + j_1)];
    }
  }
}
```


I've evaluated the correctness on V100, I think it should also work on your card P40 :)
